oak_crypto/src/encryptor.rs

```diff
@@ -0,0 +1,264 @@
+    /// Only sent in the initial request message of the session.
+    serialized_encapsulated_public_key: Option<Vec<u8>>,
+    sender_request_context: SenderRequestContext,
+    sender_response_context: SenderResponseContext,
+}
+
+impl SenderEncryptor {
+    /// Creates an HPKE crypto context by generating an new ephemeral key pair.
+    /// The `serialized_recipient_public_key` must be a NIST P-256 SEC1 encoded point public key.
+    /// <https://secg.org/sec1-v2.pdf>
+    pub fn create(serialized_recipient_public_key: &[u8]) -> anyhow::Result<Self> {
+        let (serialized_encapsulated_public_key, sender_request_context, sender_response_context) =
+            setup_base_sender(serialized_recipient_public_key, OAK_HPKE_INFO)
+                .context("couldn't create sender crypto context")?;
+        Ok(Self {
+            serialized_encapsulated_public_key: Some(serialized_encapsulated_public_key.to_vec()),
+            sender_request_context,
+            sender_response_context,
+        })
+    }
+
+    /// Encrypts `plaintext` and authenticates `associated_data` using AEAD.
+    /// Returns a [`EncryptedRequest`] proto message.
+    /// <https://datatracker.ietf.org/doc/html/rfc5116>
+    pub fn encrypt(
+        &mut self,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> anyhow::Result<EncryptedRequest> {
+        let ciphertext = self
+            .sender_request_context
+            .seal(plaintext, associated_data)
+            .context("couldn't encrypt request")?;
+        let request = EncryptedRequest {
+            encrypted_message: Some(AeadEncryptedMessage {
+                ciphertext,
+                associated_data: associated_data.to_vec(),
+            }),
+            /// Encapsulated public key is only sent in the initial request message of the session.
+            serialized_encapsulated_public_key: self.serialized_encapsulated_public_key.take(),
+        };
+        Ok(request)
+    }
+
+    /// Decrypts a [`EncryptedResponse`] proto message using AEAD.
+    /// Returns a response message plaintext and associated data.
+    /// <https://datatracker.ietf.org/doc/html/rfc5116>
+    pub fn decrypt(
+        &mut self,
+        encrypted_response: &EncryptedResponse,
+    ) -> anyhow::Result<(Vec<u8>, Vec<u8>)> {
+        let encrypted_message = encrypted_response
+            .encrypted_message
+            .as_ref()
+            .context("response doesn't contain encrypted message")?;
+        let plaintext = self
+            .sender_response_context
+            .open(
+                &encrypted_message.ciphertext,
+                &encrypted_message.associated_data,
+            )
+            .context("couldn't decrypt response")?;
+        Ok((plaintext, encrypted_message.associated_data.to_vec()))
+    }
+}
+
+enum RecipientEncryptorState {
+    Uninitialized,
+    Initialized(Box<RecipientRequestContext>, Box<RecipientResponseContext>),
```

Comment:
Why are these boxed?

Reply:
`RecipientRequestContext` has a large `inner` object, so the compiler was complaining

Reply:
Does it also complain when using `Option<CryptoContext>` rather than this custom enum?

Reply:
Done